feat(ssi): add target based filtering

betterengineering opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
* [feat(ssi): add target based configuration options](https://github.com/DataDog/datadog-agent/commit/544d1d1eecb0e31989b814f8defde41dc3b961ec) 
    * This commit adds target based configuration options to the instrumentation config.
* [feat(ssi): add target based filtering](https://github.com/DataDog/datadog-agent/commit/e73733abaae5b3936c5b48e3f90e96ed1329ffa2) 
    * This commit adds a target filter that will provide a list of tracer libraries to inject for a given pod based on a user defined set of targets.

### Motivation
As part of the [Kubernetes SSI | Workload Selection](https://docs.google.com/document/d/1Lol1ExLF1nGBe7njO6DBVkjuYAYxC1-sL2WP0rdqFRk/edit?usp=sharing) RFC, we want to allow users to have better controls over what workloads they target for Single Step Instrumentation.

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

This change is tracked entirely by unit tests. My next PR will include wiring this filter into the current injection process and will include more thorough and manual QA.

### Possible Drawbacks / Trade-offs
See the alternatives considered section of the RFC.

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->
This change is purely implementing and testing the core filtering component. My next change will include wiring this into the webhook with the existing logic. I'll add a QA tag and release note to that PR.

betterengineering committed:
feat(ssi): add target based configuration options

This commit adds target based configuration options to the
instrumentation config.

betterengineering committed:
feat(ssi): add target based filtering

This commit adds a target filter that will provide a list of tracer
libraries to inject for a given pod based on a user defined set of
targets.

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/config.go:
```diff
@@ -44,6 +46,83 @@ type InstrumentationConfig struct {
+// PodSelector is a reconstruction of the metav1.LabelSelector struct to be able to unmarshal the configuration. It
+// can be converted to a metav1.LabelSelector using the AsLabelSelector method. Full config key:
+// apm_config.instrumentation.targets[].selector
+type PodSelector struct {
```
I would love for this to be a `WorkloadSelector` with `NamespaceSelector` having the same options along with `MatchNames`.

betterengineering replied:
Adding this! The logic will be `MatchNames` OR `label.Selector`. Will take me a sec to add this in.

betterengineering replied:
Actually, we can't support this unless we add a query for the namespace on every pod start given we don't have access to the namespace metadata and only have the namespace name: https://github.com/DataDog/datadog-agent/blob/53a9dca573af4f1dd3a154f9b829cade4dc6d07e/cmd/cluster-agent/admission/server.go#L44

adel121 replied:
> Actually, we can't support this unless we add a query for the namespace on every pod start given we don't have access to the namespace metadata and only have the namespace name:

We do have namespace metadata in workloadmeta of the cluster agent. 

The namespace name is already available on the pod manifest when received by the admission controller. The admission controller runs in the cluster agent, and so does workloadmeta. The admission controller can then fetch the namespace metadata by querying the workloadmeta store.

[Here](https://github.com/DataDog/datadog-agent/blob/2910fe6cdf4ba5141e784d087b1a2cd7361ae8fd/comp/core/workloadmeta/def/component.go#L67-L69) is the right method to use.

To query for kube metadata, you need to generate kubemetadata id. We already have a function that helps you doing [this](https://github.com/DataDog/datadog-agent/blob/db518f4b0da801580e578eea3c42894a82b2ed1a/comp/core/workloadmeta/collectors/util/kube_metadata_util.go#L15-L20).

So you can retrieve namespace metadata by doing:
```
wlmStore.GetKubernetesMetadata(GenerateKubeMetadataEntityID("", "namespaces","","some-namespace"))
```

betterengineering replied:
If it's ok with the both of you, I will follow up on this in [INPLAT-466](https://datadoghq.atlassian.net/browse/INPLAT-466)

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/config.go:
```diff
@@ -60,6 +139,16 @@ func NewInstrumentationConfig(datadogConfig config.Component) (*InstrumentationC
+
+	// Ensure both library versions and targets are not set together.
+	if len(cfg.LibVersions) > 0 && len(cfg.Targets) > 0 {
+		return nil, fmt.Errorf("apm.instrumentation.lib_versions and apm.instrumentation.targets are mutually exclusive and cannot be set together")
```
🤔 What are your thoughts on these being the default versions we'd use for a target if none were specified at the `Target` level?

betterengineering replied:
I suspect it will be confusing to the user and be less explicit, especially given `lib_versions` is used today to "target" workloads.

betterengineering replied:
After thinking about this more, I think config values should NOT change meaning. For example, in the targets based selection, we have decided to keep `disabled_namespaces` and that's ok given the meaning doesn't change. In the case of using lib_versions to determine the default versions, it would be a behavior change from what it means today.

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter.go:
```diff
@@ -0,0 +1,116 @@
+		}
+
+		// If the namespace and pod selector match, return the libraries to inject.
+		return target.libVersions
```
Per my prior comment on default libversions, if this is `nil` we should consider having the default ones if they are specified.

betterengineering replied:
Spoke offline, we decided that we should maintain the existing behavior in that if no target matches, the user gets no tracers.

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter_test.go:
```diff
@@ -0,0 +1,132 @@
+			libList := f.Filter(test.in)
+
+			// Validate the output.
+			languages := convertLibList(libList)
```
Because our targets have potential library versions, we should add those here in the tests as well instead of only matching the languages.

betterengineering replied:
Adding!

betterengineering committed:
Address linter

betterengineering committed:
update filter test to include full lib version

This commit updates the filter test to include the full lib version,
which is more cumbersome but it's a stronger test.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/config.go:
```diff
@@ -44,6 +46,83 @@ type InstrumentationConfig struct {
+// Target is a rule to apply the auto instrumentation to a specific workload using the pod and namespace selectors.
+// Full config key: apm_config.instrumentation.targets to get the list of targets.
+type Target struct {
+	// Name is the name of the target. It will be appended to the pod annotations to identify the target that was used.
```
> It will be appended to the pod annotations to identify the target that was used.

What if the target is modified, while the name is kept unchanged? Previously mutated pods will still show the same target name in the annotations, which might appear conflicting when compared to the new target config (i.e. the new target implies that the pod should not be instrumented, and yet it appears in the pod annotations).

betterengineering replied:
This is a good call out. What if we also add a timestamp? So `Foo` target at 10am was selected for this workload. That should provide additional clarity for introspection. We could also create a hash of the rule and also apply that.

betterengineering replied:
Will follow up on this in [INPLAT-465](https://datadoghq.atlassian.net/browse/INPLAT-465)

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/config.go:
```diff
@@ -44,6 +46,83 @@ type InstrumentationConfig struct {
+	// Selector is the pod selector to match the pods to apply the auto instrumentation to. It will be used in
+	// conjunction with the NamespaceSelector to match the pods. Full config key:
+	// apm_config.instrumentation.targets[].selector
+	Selector PodSelector `mapstructure:"selector"`
```
Nit:

Maybe rename this property to `PodSelector`? 

Seems confusing at a first sight to have `Selector` and `NamespaceSelector`

betterengineering replied:
Will update!

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/config.go:
```diff
@@ -44,6 +46,83 @@ type InstrumentationConfig struct {
+
+// PodSelector is a reconstruction of the metav1.LabelSelector struct to be able to unmarshal the configuration. It
+// can be converted to a metav1.LabelSelector using the AsLabelSelector method. Full config key:
+// apm_config.instrumentation.targets[].selector
```
nit:

> Full config key:
// apm_config.instrumentation.targets[].selector

I think this doesn't fit in here.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/config.go:
```diff
@@ -44,6 +46,83 @@ type InstrumentationConfig struct {
+// PodSelector is a reconstruction of the metav1.LabelSelector struct to be able to unmarshal the configuration. It
+// can be converted to a metav1.LabelSelector using the AsLabelSelector method. Full config key:
+// apm_config.instrumentation.targets[].selector
+type PodSelector struct {
```
Question: 

I understand from the comments that `PodSelector` is a reconstruction of `metav1.LabelSelector`. However, I don't see any key difference in the implementation. Why can't we directly use `metav1.LabelSelector` and avoid having to implement `AsLabelSelector()` for example.

betterengineering replied:
Good question! In order to read from the viper config, I needed to add the `mapstructure` struct tags. I'm unaware of a way to modify a third party's struct tags in our code, but if there is a way, that would be handy.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/config.go:
```diff
@@ -44,6 +46,83 @@ type InstrumentationConfig struct {
+
+// PodSelectorMatchExpression is a reconstruction of the metav1.LabelSelectorRequirement struct to be able to unmarshal
+// the configuration. Full config key: apm_config.instrumentation.targets[].selector.matchExpressions
+type PodSelectorMatchExpression struct {
```
same question. Why not directly use `metav1.LabelSelectorRequirement` type instead of reconstructing it?

betterengineering replied:
We need the mapstructure struct tags to read from the viper config.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/config.go:
```diff
@@ -44,6 +46,83 @@ type InstrumentationConfig struct {
+type NamespaceSelector struct {
+	// MatchNames is a list of namespace names to match. If empty, all namespaces are matched. Full config key:
+	// apm_config.instrumentation.targets[].namespaceSelector.matchNames
+	MatchNames []string `mapstructure:"matchNames"`
```
The native kubernetes `NamespaceSelector` doesn't support matching by names, only matching by labels is supported. (namespace selector is a normal label selector).

So this `NamespaceSelector` struct is not aligned with the one used by kubernetes. 

It can still work, but I think there can be 2 issues with this:
* From a k8s user POV, when I see a config named `NamespaceSelector`, I will intuitively use it as I am accustomed to do with kubernetes label selectors. (Unless I am wise enough to read the documentation thoroughly).
* Using a selector that is not aligned with kubernetes defined one will make it difficult (or maybe impossible) to reuse this as a webhook object/namespace selector.

More context about the second point: 
Upon registration of the admission controller webhook, we provide `ObjectSelector` and `NamespaceSelector`. For auto instrumentation webhook, this is defined [here](https://github.com/DataDog/datadog-agent/blob/ff5ef16b8bf9208a4d094b76ff75f6e0cb160d9d/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go#L129-L133). The logic creating the object and namespace selectors can be found [here](https://github.com/DataDog/datadog-agent/blob/ff5ef16b8bf9208a4d094b76ff75f6e0cb160d9d/pkg/clusteragent/admission/common/label_selectors.go#L17-L51). You can see there is a comment saying `// Accept all, ignore pods if they're explicitly filtered-out`. This can be improved a lot, but due to dependency constraints across webhooks, the code needs some refactoring. The issue with the existing code is that even if a pod is definitely not to be instrumented, it will still go through the webhook, and might fail to start if the DCA is not available to process the pod mutation request (and we've already got multiple support tickets about this).

All this is to say that the selectors that are added in this PR will make it difficult (or nearly impossible) to reuse them in the webhook filters on registration as long as they don't perfectly align with the kubernetes label selector construct. 

This raises the question of whether matching targets by namespace name is strictly necessary or if we can rely on namespace labels instead. (there is a tradeoff between performance and ease-of-use)

betterengineering replied:
When @stanistan and I were going back of forth on the design, this came up a few times. We eventually decided that `MatchNames` for namespaces was more ergonomic and provided a more clear path for folks using `enabledNamespaces` today.

For the first issue, there is some precedence for the `MatchNames` field, as the Prometheus Operator has a MatchNames](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/running-exporters.md#generic-servicemonitor-example). So it's not entirely out of the ordinary.

For the second issue, we don't use the webhook namespace selector today and instead process all pods, so it would be an optimization. Even without MatchNames, we would have to calculate the potential namespaces using the targets list, so we would already have to compute a namespace list. The last thing to call out is that only MatchNames OR (MatchLabels/MatchExpressions) would be allowed in the config. So we would still be able to compute a list of namespaces to operate on if we wanted to implement this down the line.

stanistan replied:
Agree with @adel121 about having this be a standard selector + the ability to do `MatchNames` which in the context of a namespace is equivalent given the namespace label `kubernetes.io/metadata.name` is automatically applied: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#automatic-labelling

I also thought this was a discussion for the RFC (but maybe misremembering). Is the plan to add namespace label selection in a follow up?

davidor replied:
I agree with @adel121 's point about using labels in the namespace selector.

Apart from what he mentioned, I agree that filtering by name is simple to understand, but I think it lacks flexibility. If some rules change, adding or removing labels is easy, whereas renaming a namespace is not possible.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter.go:
```diff
@@ -0,0 +1,116 @@
+	libVersions       []libInfo
+}
+
+// NewTargetFilter creates a new TargetFilter from a list of targets and disabled namespaces. We convert the targets
+// to a more efficient internal format for quick lookups.
```
Question:

This makes since if the user has configured `disabledNamespaces`. But what if they configure `enabledNamespaces`? How can this function be used then?

betterengineering replied:
It is an error condition to have both `enabledNamespaces` and `targets` set (see `NewInstruementationConfig()` in this PR). The reason is because it would be unclear if `enabledNamespaces` took precedence or if the `targets` took precedence, and only allowing one clears up that potential for confusion.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter.go:
```diff
@@ -0,0 +1,116 @@
+	containerRegistry  string
+}
+
+// targetInternal is the struct we use to conver the config based target into
```
```suggestion
// targetInternal is the struct we use to convert the config based target into
```

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter.go:
```diff
@@ -0,0 +1,116 @@
+// TargetFilter filters pods based on a set of targeting rules.
+type TargetFilter struct {
+	targets            []targetInternal
+	disabledNamespaces map[string]bool
```
Question: why is `disabledNamespaces` defined on the level of `TargetFilter` while `enabledNamespaces` is defined on `targetInternal` ?

betterengineering committed:
Update pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter.go

Co-authored-by: Adel Haj Hassan <41540817+adel121@users.noreply.github.com>

betterengineering committed:
rename selector to podSelector to make it more clear.

This commit renames the selector to podSelector as it makes it clearer
what the selector is selecting.

betterengineering committed:
Merge branch 'main' into mark.spicer/INPLAT-461-add-target-filtering

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter.go:
```diff
@@ -0,0 +1,115 @@
+)
+
+// TargetFilter filters pods based on a set of targeting rules.
+type TargetFilter struct {
```
Why is this a public type? If I'm reading right it has no public methods or fields.

betterengineering replied:
Add a TODO for this

betterengineering commented:
/merge
